# Acquire Exercises

In [80]:
from bs4 import BeautifulSoup
import re
import pandas as pd

import requests

# 1. Codeup Blog Articles

Visit Codeup's Blog and record the urls for at least 5 distinct blog posts. For each post, you should scrape at least the post's title and content.

Encapsulate your work in a function named get_blog_articles that will return a list of dictionaries, with each dictionary representing one article. The shape of each dictionary should look like this:

```python
{
    'title': 'the title of the article',
    'content': 'the full text content of the article'
}
```
Plus any additional properties you think might be helpful.

Bonus: Scrape the text of all the articles linked on codeup's blog page.

In [2]:
# url
url='https://codeup.com/blog/'

In [3]:
# create soup
soup = BeautifulSoup(requests.get(url).content, 'html.parser')

In [4]:
# see soup
soup

<html>
<head><title>403 Forbidden</title></head>
<body>
<center><h1>403 Forbidden</h1></center>
<hr/><center>nginx</center>
</body>
</html>

In [5]:
# request text
html_text =requests.get(url).text

In [6]:
print(html_text)

<html>
<head><title>403 Forbidden</title></head>
<body>
<center><h1>403 Forbidden</h1></center>
<hr><center>nginx</center>
</body>
</html>



I got code 403; trying to figure a way to acess the information.

New attempt follows:

In [7]:
#Define headers
headers = {'User-Agent': 'Codeup Data Science'}

In [8]:
#Make a request
response = requests.get('https://codeup.com/data-science/become-a-data-scientist/', headers=headers)

In [9]:
#Create the soup and investigate
soup = BeautifulSoup(response.content, 'html.parser')


soup

In [10]:
title= soup.find('h1')
title.text

'Become a Data Scientist in 6 Months!'

In [11]:
#Access the date published
publish_date = soup.find('span', class_='published')
publish_date.text

'Jan 5, 2023'

In [12]:
#Access the article content
content = soup.find('div', class_='entry-content')
content.text

'\nAre you feeling unfulfilled in your work but want to avoid returning to the traditional educational route? Codeup can help! Starting over as a professional is daunting and not always ideal. Codeup can help you go from a career you are bored with, to a job that excites you in just 6 months!\nHere’s how…\nData Science Program\nDuring our 20-week program, you will have the opportunity to take your career to new heights with data science being one of the most needed jobs in tech.\nYou’ll gather data, then clean it, explore it for trends, and apply machine learning models to make predictions.\nUpon completing this program, you will know how to turn insights into actionable recommendations. You’ll be a huge asset to any company, having all the technical skills to become a data scientist with projects upon projects of experience under your belt.\nCodeup\nA common reason individuals opt not to change their careers is fear it is too late. Codeup has crafted a program that will guide you thro

In [13]:
article_links = soup.find_all('a', 'href')
article_links

[]

In [14]:
# Get the "a" tag
element = soup.find('a')
# Get the attribute value
data = element.get('href')
data

'https://codeup.com/'

In [15]:
# Get the "a" tag
element = soup.find_all('a')
# Get the attribute value
element


[<a href="https://codeup.com/">Home</a>,
 <a href="https://codeup.com/program/cloud-adminsitration/">Cloud Administration</a>,
 <a href="https://codeup.com/program/full-stack-web-development/">Full Stack Web Development</a>,
 <a href="https://codeup.com/program/data-science/">Data Science</a>,
 <a href="/events/">Workshops</a>,
 <a href="/san-antonio-events/">San Antonio</a>,
 <a href="/dallas-events/">Dallas</a>,
 <a href="https://codeup.com/financial-aid/">Financial Aid</a>,
 <a href="https://codeup.com/veterans/">Military</a>,
 <a href="https://codeup.com/hire-tech-talent/">Hire Tech Talent</a>,
 <a href="https://alumni.codeup.com/">Alumni</a>,
 <a href="https://codeup.com/resources/">Resources</a>,
 <a href="/my-story/">Student Reviews</a>,
 <a href="https://codeup.com/blog/">Blog</a>,
 <a href="https://codeup.com/frequently-asked-questions/">Common Questions</a>,
 <a href="https://codeup.com/podcast/">Hire Tech Podcast</a>,
 <a href="https://codeup.com/apply-now/">Apply Now</a>,
 

In [16]:
tags = soup.find_all("li", "search-result")
tags

[]

In [17]:
tag = soup.find("div", id="bar")
tag

In [18]:
tags = soup.select("#search-results .external-links")
tags

[]

In [19]:
anchor_href = soup.find("a")["href"]
anchor_href

'https://codeup.com/'

In [20]:
link_list=[]
for link in soup.find_all('a'):
    link_list.append(link.get('href'))


In [21]:
link_list=[]
for link in soup.find_all('a'):
    pattern = re.compile(r'(codeup.com\/|www.|twitter.|alumni.)[\w\/-]+|(.com|.org)[\w\/-?]+|codeup.com')
    
    matches = pattern.finditer(link.get('href'))
    print(matches)

TypeError: expected string or bytes-like object

In [22]:
blog_url_list= ['https://codeup.com/data-science/become-a-data-scientist/',
                'https://codeup.com/employers/hiring-tech-talent/',
               'https://codeup.com/cloud-administration/cap-funding-options/'
               'https://codeup.com/dallas-info/it-professionals-dallas/',
               'https://codeup.com/codeup-news/codeup-voted-1-technical-school-in-dfw/',
               'https://codeup.com/tips-for-prospective-students/financing/codeups-scholarships/']

In [23]:
def basic_info(url):
    blog_info={"title","publish_date","content"}
    headers = {'User-Agent': 'Codeup Data Science'}
    response = requests.get(url, headers=headers)
    
    #find the title
    soup = BeautifulSoup(response.content, 'html.parser')
    title = soup.find('h1')
    
    
    #Access the date published
    publish_date = soup.find('span', class_='published')
 
    
    #Access the article content
    content = soup.find('div', class_='entry-content')

    output = {
        'title': title.text,
        'publish_date': publish_date.text,
        'content': content.text
     }
    return output

In [24]:
blog_info= []
for i in blog_url_list:
    output = basic_info(i)
    blog_info.append(output)
blog_info

[{'title': 'Become a Data Scientist in 6 Months!',
  'publish_date': 'Jan 5, 2023',
  'content': '\nAre you feeling unfulfilled in your work but want to avoid returning to the traditional educational route? Codeup can help! Starting over as a professional is daunting and not always ideal. Codeup can help you go from a career you are bored with, to a job that excites you in just 6 months!\nHere’s how…\nData Science Program\nDuring our 20-week program, you will have the opportunity to take your career to new heights with data science being one of the most needed jobs in tech.\nYou’ll gather data, then clean it, explore it for trends, and apply machine learning models to make predictions.\nUpon completing this program, you will know how to turn insights into actionable recommendations. You’ll be a huge asset to any company, having all the technical skills to become a data scientist with projects upon projects of experience under your belt.\nCodeup\nA common reason individuals opt not to c

# 2. News Articles

We will now be scraping text data from inshorts, a website that provides a brief overview of many different topics.

Write a function that scrapes the news articles for the following topics:

* Business
* Sports
* Technology
* Entertainment

The end product of this should be a function named get_news_articles that returns a list of dictionaries, where each dictionary has this shape:
``` python

{
    'title': 'The article title',
    'content': 'The article content',
    'category': 'business' # for example
}
```
Hints:

a. Start by inspecting the website in your browser. Figure out which elements will be useful.

b. Start by creating a function that handles a single article and produces a dictionary like the one above.

c. Next create a function that will find all the articles on a single page and call the function you created in the last step for every article on the page.

d. Now create a function that will use the previous two functions to scrape the articles from all the pages that you need, and do any additional processing that needs to be done.

In [114]:
# set url
url = 'https://inshorts.com/en/read/technology'


In [115]:
# get content
response = requests.get(url)
response = response.content

In [141]:
# to get html code, nicer looking
soup = BeautifulSoup(response,'html.parser' )
#soup

In [133]:
div = soup.find_all('div', class_ = 'card stack')

In [162]:
span = soup.find_all('span')

In [164]:
itemprop = soup.find_all('itemprop')

In [165]:
itemprop

[]

In [157]:
author = soup.find('span',class_='author').text
date = soup.find('span',class_='date').text

title = soup.find('span itemprop',class_='headline')
author,date, title

('Hiral Goyal', '07 Jan', None)

In [132]:
article_thumb = div['class']="card stack"
article_thumb

'card stack'

In [ ]:
articles = div.find_all('', class_='product_pod'

# 3. Bonus: cache the data

Write your code such that the acquired data is saved locally in some form or fashion. Your functions that retrieve the data should prefer to read the local data instead of having to make all the requests everytime the function is called. Include a boolean flag in the functions to allow the data to be acquired "fresh" from the actual sources (re-writing your local cache).

# 4. Extra Practice: books

In [25]:
# set url
url ='http://books.toscrape.com'

In [36]:
# response of 200 means it was sucessful
response = requests.get(url)
response

<Response [200]>

In [37]:
response_content = response.content

In [42]:
# remove hashtag below to see content
#response_content

In [41]:
# to get html code, nicer looking
soup = BeautifulSoup(response_content,'html.parser' )


In [108]:
# remove hashtag below to see soup (html content)
#soup

In [45]:
ol = soup.find('ol')

In [46]:
articles = ol.find_all('article', class_='product_pod')

In [50]:
# Remove hastab below to see list of articles
articles[0]

<article class="product_pod">
<div class="image_container">
<a href="catalogue/a-light-in-the-attic_1000/index.html"><img alt="A Light in the Attic" class="thumbnail" src="media/cache/2c/da/2cdad67c44b002e7ead0cc35693c0e8b.jpg"/></a>
</div>
<p class="star-rating Three">
<i class="icon-star"></i>
<i class="icon-star"></i>
<i class="icon-star"></i>
<i class="icon-star"></i>
<i class="icon-star"></i>
</p>
<h3><a href="catalogue/a-light-in-the-attic_1000/index.html" title="A Light in the Attic">A Light in the ...</a></h3>
<div class="product_price">
<p class="price_color">£51.77</p>
<p class="instock availability">
<i class="icon-ok"></i>
    
        In stock
    
</p>
<form>
<button class="btn btn-primary btn-block" data-loading-text="Adding..." type="submit">Add to basket</button>
</form>
</div>
</article>

In [49]:
# to find the title of the articles
for article in articles:
    image = article.find('img')
    title = image.attrs['alt']
    print(title)

A Light in the Attic
Tipping the Velvet
Soumission
Sharp Objects
Sapiens: A Brief History of Humankind
The Requiem Red
The Dirty Little Secrets of Getting Your Dream Job
The Coming Woman: A Novel Based on the Life of the Infamous Feminist, Victoria Woodhull
The Boys in the Boat: Nine Americans and Their Epic Quest for Gold at the 1936 Berlin Olympics
The Black Maria
Starving Hearts (Triangular Trade Trilogy, #1)
Shakespeare's Sonnets
Set Me Free
Scott Pilgrim's Precious Little Life (Scott Pilgrim #1)
Rip it Up and Start Again
Our Band Could Be Your Life: Scenes from the American Indie Underground, 1981-1991
Olio
Mesaerion: The Best Science Fiction Stories 1800-1849
Libertarianism for Beginners
It's Only the Himalayas


In [61]:
# to find the star of the articles
for article in articles:
    star = article.find('p')
    rating = star['class'][1]
    print(rating)

Three
One
One
Four
Five
One
Four
Three
Four
One
Two
Four
Five
Five
Five
Three
One
One
Two
Two


In [72]:
# to find the price of the articles
for article in articles:
    price = article.find('p', class_="price_color").text
    # to convert text to float without money symbol
    price = float(price[1:])
    print(price)

51.77
53.74
50.1
47.82
54.23
22.65
33.34
17.93
22.6
52.15
13.99
20.66
17.46
52.29
35.02
57.25
23.88
37.59
51.33
45.17


In [95]:
# get information of one page in book website
books =[]
for article in articles:
    image = article.find('img')
    title = image.attrs['alt']
    star = article.find('p')
    rating = star['class'][1]
    price = article.find('p', class_="price_color").text
    # to convert text to float without money symbol
    price = float(price[1:])
    books.append([title,price,rating])
books_df = pd.DataFrame(books, columns = ['title','price','stars'])
books_df

title  price  stars
0                                A Light in the Attic  51.77  Three
1                                  Tipping the Velvet  53.74    One
2                                          Soumission  50.10    One
3                                       Sharp Objects  47.82   Four
4               Sapiens: A Brief History of Humankind  54.23   Five
5                                     The Requiem Red  22.65    One
6   The Dirty Little Secrets of Getting Your Dream...  33.34   Four
7   The Coming Woman: A Novel Based on the Life of...  17.93  Three
8   The Boys in the Boat: Nine Americans and Their...  22.60   Four
9                                     The Black Maria  52.15    One
10     Starving Hearts (Triangular Trade Trilogy, #1)  13.99    Two
11                              Shakespeare's Sonnets  20.66   Four
12                                        Set Me Free  17.46   Five
13  Scott Pilgrim's Precious Little Life (Scott Pi...  52.29   Five
14                          Rip it Up and Start Again  35.02   Five
15  Our Band Could Be Your Life: Scenes from the A...  57.25  Three
16                                               Olio  23.88    One
17  Mesaerion: The Best Science Fiction Stories 18...  37.59    One
18                       Libertarianism for Beginners  51.33    Two
19                            It's Only the Himalayas  45.17    Two

In [98]:
# get book information of all the pages in website
books=[]
for i in range(1,51):
    
    url = f"http://books.toscrape.com/catalogue/page-{i}.html"
    response = requests.get(url)
    response = response.content
    soup = BeautifulSoup(response,'html.parser')
    ol = soup.find('ol')
    artticles = ol.find_all('article', class_ = 'product_pod')



    for article in articles:
        image = article.find('img')
        title = image.attrs['alt']
        star = article.find('p')
        rating = star['class'][1]
        price = article.find('p', class_="price_color").text
        # to convert text to float without money symbol
        price = float(price[1:])
        books.append([title,price,rating])
books_df = pd.DataFrame(books, columns = ['title','price','stars'])
books_df

title  price  stars
0                                 A Light in the Attic  51.77  Three
1                                   Tipping the Velvet  53.74    One
2                                           Soumission  50.10    One
3                                        Sharp Objects  47.82   Four
4                Sapiens: A Brief History of Humankind  54.23   Five
..                                                 ...    ...    ...
995  Our Band Could Be Your Life: Scenes from the A...  57.25  Three
996                                               Olio  23.88    One
997  Mesaerion: The Best Science Fiction Stories 18...  37.59    One
998                       Libertarianism for Beginners  51.33    Two
999                            It's Only the Himalayas  45.17    Two

[1000 rows x 3 columns]